includes/classes/Feature/GeoLocation.php

```diff
@@ -0,0 +1,561 @@
+		add_filter( 'ep_formatted_args', [ $this, 'formatted_args' ], 10, 2 );
+		add_action( 'pre_get_posts', [ $this, 'change_query' ] );
+
+		add_action( 'init', [ $this, 'register_block' ] );
+		add_action( 'wp_enqueue_scripts', [ $this, 'enqueue_assets' ] );
+		add_action( 'parse_request', [ $this, 'maybe_change_cookie' ] );
+	}
+
+	/**
+	 * Enqueue admin scripts
+	 *
+	 * @return void
+	 */
+	public function admin_scripts(): void {
+		wp_enqueue_script(
+			'ep_geo_location_script',
+			ELASTICPRESS_LABS_URL . 'dist/js/geo-location-script.js',
+			Utils\get_asset_info( 'geo-location-script', 'dependencies' ),
+			Utils\get_asset_info( 'geo-location-script', 'version' ),
+			true
+		);
+
+		wp_set_script_translations( 'ep_geo_location_script', 'elasticpress-labs' );
+
+		$google_maps_api_key = $this->get_setting( 'google_maps_api_key' );
+
+		wp_localize_script(
+			'ep_geo_location_script',
+			'epGeoLocation',
+			[
+				'has_map_key'      => ! empty( $google_maps_api_key ),
+				'is_external_meta' => has_filter( 'ep_geo_location_pre_geo_points' ),
+			]
+		);
+
+		if ( ! empty( $google_maps_api_key ) ) {
+			$google_places_api_url = add_query_arg(
+				[
+					'key'       => $google_maps_api_key,
+					'libraries' => 'places',
+				],
+				'https://maps.googleapis.com/maps/api/js'
+			);
+
+			wp_enqueue_script(
+				'google-places-api',
+				$google_places_api_url,
+				[],
+				ELASTICPRESS_LABS_VERSION,
+				true
+			);
+		}
+	}
+
+	/**
+	 * Set the `settings_schema` attribute
+	 *
+	 * @return void
+	 */
+	public function set_settings_schema(): void {
+		$this->settings_schema = [
+			[
+				'default' => '',
+				'help'    => __( 'Providing a Google Maps API key enables an autocomplete address field that automatically fetches the latitude and longitude.', 'elasticpress-labs' ),
+				'key'     => 'google_maps_api_key',
+				'label'   => __( 'Google Maps API Key', 'elasticpress-labs' ),
+				'type'    => 'text',
+			],
+		];
+	}
+
+	/**
+	 * Register meta fields
+	 *
+	 * @return void
+	 */
+	public function register_meta(): void {
+		register_post_meta(
+			'',
+			'ep_latitude',
+			[
+				'type'         => 'number',
+				'description'  => esc_html__( 'Latitude', 'elasticpress-labs' ),
+				'single'       => true,
+				'show_in_rest' => true,
+			]
+		);
+
+		register_post_meta(
+			'',
+			'ep_longitude',
+			[
+				'type'         => 'number',
+				'description'  => esc_html__( 'Longitude', 'elasticpress-labs' ),
+				'single'       => true,
+				'show_in_rest' => true,
+			]
+		);
+
+		register_post_meta(
+			'',
+			'ep_address',
+			[
+				'type'         => 'string',
+				'description'  => esc_html__( 'Address', 'elasticpress-labs' ),
+				'single'       => true,
+				'show_in_rest' => true,
+			]
+		);
+	}
+
+	/**
+	 * Add mapping for geo_point.
+	 *
+	 * @param array $mapping Mapping.
+	 * @return array Mapping.
+	 */
+	public function add_mapping( $mapping ): array {
+		$mapping['mappings']['properties']['geo_point'] = [
```

Comment:
@burhandodhy this will only work on ES 7+. If geo_point is not available on ES 5.2 we should not allow people to activate it. If it is, we'll need something like

```php
		if ( version_compare( (string) Elasticsearch::factory()->get_elasticsearch_version(), '7.0', '<' ) ) {
			$mapping_properties = &$mapping['mappings']['post']['properties'];
		} else {
			$mapping_properties = &$mapping['mappings']['properties'];
		}

		$mapping_properties['geo_point'] = [
```